kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java

```diff
@@ -112,124 +96,26 @@ public BigQuerySinkTask() {
     this.testBigQuery = testBigQuery;
   }
 
-  // Called asynchronously from TableWriter.call(); synchronization required on context
-  private void updateAllPartitions(String topic, Map<TopicPartition, OffsetAndMetadata> offsets) {
-    Map<TopicPartition, Long> topicOffsets = new HashMap<>();
-    for (Map.Entry<TopicPartition, OffsetAndMetadata> topicPartitionOffset : offsets.entrySet()) {
-      TopicPartition topicPartition = topicPartitionOffset.getKey();
-      if (topicPartition.topic().equals(topic)) {
-        topicOffsets.put(topicPartition, topicPartitionOffset.getValue().offset());
-      }
-    }
-    if (!topicOffsets.isEmpty()) {
-      synchronized (contextLock) {
-        context.offset(topicOffsets);
-      }
-    }
-  }
-
-  private class TableWriter implements Callable<Void> {
-    private final PartitionedTableId partitionedTableId;
-    private final List<RowToInsert> rows;
-    private final Map<TopicPartition, OffsetAndMetadata> offsets;
-    private final String topic;
-    private final BatchWriter<RowToInsert> batchWriter;
-
-    public TableWriter(PartitionedTableId partitionedTableId,
-                       List<RowToInsert> rows,
-                       Map<TopicPartition, OffsetAndMetadata> offsets,
-                       String topic) {
-      this.partitionedTableId = partitionedTableId;
-      this.rows = rows;
-      this.offsets = offsets;
-      this.topic = topic;
-      this.batchWriter = batchWriterManager.getBatchWriter(partitionedTableId.getBaseTableId());
+  @Override
+  public void flush(Map<TopicPartition, OffsetAndMetadata> offsets) {
+    try {
+      executor.awaitCurrentTasks();
+    } catch (InterruptedException err) {
+      throw new ConnectException("Interrupted while waiting for write tasks to complete.", err);
     }
+    updateOffsets(offsets);
 
-    @Override
-    public Void call() throws InterruptedException {
-      batchWriter.writeAll(partitionedTableId, rows, topic);
-      updateAllPartitions(baseTableIdsToTopics.get(partitionedTableId.getBaseTableId()), offsets);
-      return null;
-    }
+    topicPartitionManager.resumeAll();
   }
 
   /**
-   * A class for keeping track of the BatchWriters for each base table.
+   * This really doesn't do much and I'm not totally clear on whether or not I need it.
```

Comment:
Very late to the game here, but since I found this via https://issues.apache.org/jira/browse/KAFKA-6148 figured I'd mention that this code should not, in fact, be needed. If you actually flush all data in `flush()` and do not throw an exception, the correct offsets will be automatically committed. You only need to use `context.offset` if you want to fully manage offsets yourself.

Reply:
I could be off-base here but I think the reason we initially included manual updating of offsets was to help push things forward in the event of a failed `flush()` due to an excessive volume of data. If too many records were meant to be written in a call to `flush()` and an exception ended up being thrown, the manual updates would prevent the next call to `flush()` from doing the exact same thing and trapping the sink task in an endless loop. If there are already steps in place for the Connect framework to handle this case automatically then the code block is probably obsolete and can be removed safely.

Reply:
If an exception is thrown from flush, the connector will still continue from where it was, it would just be unable to commit offsets. You can see how that is handled [here](https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java#L343-L359) though now `preCommit()` wraps `flush()` by default, so that's the relevant call. Note that we don't modify the consumer in the event of failure, we just return from the call, so the consumer continues where it left off. Perhaps there are some javadoc updates in the SinkTask interface that would help clarify this.

For this connector there's probably a better fix [KIP-89](https://cwiki.apache.org/confluence/display/KAFKA/KIP-89%3A+Allow+sink+connectors+to+decouple+flush+and+offset+commit) allows the sink task better control over how offset commits work, allowing it to, e.g., not immediately flush data but just return what safe offsets to commit are. You have to be a bit careful to not allow your backlog to grow indefinitely (i.e. don't consistently accept data from Kafka faster than you can write it to your sink), but this allows you to just very quickly return from `preCommit` and never have to fully flush the pipeline.

In any case, we tracked down the issue to actually be that the build process for this connector is incorrectly including `connect-api` instead of marking it provided, which in turn pulls in a copy of `kafka-clients` which doesn't play well with the new classloader isolation. We're going to make a fix to Connect so that even if you include the jar we'll load the clients classes from the right place, but Connectors shouldn't be including a copy of either jar since that can make them only work with the specific version of Connect they were compiled with when using loading of plugins via CLASSPATH.

I filed https://github.com/wepay/kafka-connect-bigquery/pull/100 which should fix the issue except that debezium has the same problem. I don't know gradle well enough to deal with excluding all transitive dependencies to make this easily fixed without also fixing the debezium issue.